backend/provisioner/inmem_provisioner.go

```diff
@@ -0,0 +1,157 @@
+package provisioner
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync/atomic"
+
+	"connectrpc.com/connect"
+	"github.com/google/uuid"
+	"github.com/puzpuzpuz/xsync/v3"
+
+	ftlv1 "github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1"
+	"github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1beta1/provisioner"
+	"github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1beta1/provisioner/provisionerconnect"
+	"github.com/TBD54566975/ftl/internal/log"
+)
+
+type inMemProvisioningTask struct {
+	steps []*inMemProvisioningStep
+}
+
+func (t *inMemProvisioningTask) Done() (bool, error) {
+	done := true
+	for _, step := range t.steps {
+		if !step.Done.Load() {
+			done = false
+		}
+		if step.Err != nil {
+			return false, step.Err
+		}
+	}
+	return done, nil
+}
+
+type inMemProvisioningStep struct {
+	Resource *provisioner.Resource
+	Err      error
+	Done     atomic.Bool
+}
+
+// InMemResourceProvisionerFn is a function that provisions a resource
+type InMemResourceProvisionerFn func(context.Context, *provisioner.ResourceContext, string, string) (*provisioner.Resource, error)
+
+// InMemProvisioner for running an in memory provisioner, constructing all resources concurrently
+//
+// It spawns a separate goroutine for each resource to be provisioned, and
+// finishes the task when all resources are provisioned or an error occurs.
+type InMemProvisioner struct {
+	running  *xsync.MapOf[string, *inMemProvisioningTask]
+	handlers map[ResourceType]InMemResourceProvisionerFn
+}
+
+func NewEmbeddedProvisioner(handlers map[ResourceType]InMemResourceProvisionerFn) *InMemProvisioner {
+	return &InMemProvisioner{
+		running:  xsync.NewMapOf[string, *inMemProvisioningTask](),
+		handlers: handlers,
+	}
+}
+
+var _ provisionerconnect.ProvisionerPluginServiceClient = (*InMemProvisioner)(nil)
+
+func (d *InMemProvisioner) Ping(context.Context, *connect.Request[ftlv1.PingRequest]) (*connect.Response[ftlv1.PingResponse], error) {
+	return &connect.Response[ftlv1.PingResponse]{}, nil
+}
+
+func (d *InMemProvisioner) Plan(context.Context, *connect.Request[provisioner.PlanRequest]) (*connect.Response[provisioner.PlanResponse], error) {
+	panic("unimplemented")
+}
+
+func (d *InMemProvisioner) Provision(ctx context.Context, req *connect.Request[provisioner.ProvisionRequest]) (*connect.Response[provisioner.ProvisionResponse], error) {
```

Comment:
It doesn't feel ideal for the gRPC service interface to also be the abstraction boundary for different implementations? Dealing with protos at the application logic layer is always pretty gross :(